File: server/src/battle/DeskCardMgr.java
package battle;

import model.user.card.Card;
import bitzero.server.entities.User;
import config.user.CardConfig;
import model.user.card.CardModel;
import model.user.card.CardService;

public class DeskCardMgr {
    public int index;
    public int numberOfCard;
    public Card[] listCard;
    public boolean[] isShowed;
    
    public DeskCardMgr(User user, int[] listDeskCard) {
        index = -1;
        numberOfCard = CardConfig.numberCardUsing;
        listCard = new Card[numberOfCard];
        isShowed = new boolean[numberOfCard];

        // Load list card from user
        CardModel cardModel = CardService.getInstance().getModelFromCache(user);
        for (int i = 0; i < cardModel.listCardUsing.length; i++) {
            listCard[i] = cardModel.listCard[listDeskCard[i]];
        }

        for (int i = 0; i < numberOfCard; i++) {
            isShowed[i] = false;
        }

        for (int i = 0; i < 4; i++) {
            getNextIndex();
        }
    }

    public int getNextIndex() {
        do { 
            index = (index + 1) % numberOfCard;
        } while (isShowed[index]);
        isShowed[index] = true;
        return index;
    }

    public void hideCard(int index) {
        isShowed[index] = false;
    }
}

File: client/src/modules/battle/logic/DeskCardMgr.js
let DeskCardMgr = cc.Class.extend({

    ctor: function (listDeskCard, listDeskCardLevel) {
        this.index = -1
        this.numberOfCard = 8
        this.showed = []
        this.listCard = []

        for (let i = 0; i < listDeskCard.length; i++) {
            this.listCard.push(new Card({
                id: listDeskCard[i],
                level: listDeskCardLevel[i],
            }))
        }

        for (var i = 0; i < this.numberOfCard; i++) {
            this.showed[i] = false
        }
    },

    getNextIndex: function () {
        do {
            this.index = (this.index + 1) % this.numberOfCard
        } while (this.showed[this.index])
        this.showed[this.index] = true
        return this.index
    },

    getNextIndexForCardNext: function () {
        var tmpIndex = this.index
        do {
            tmpIndex = (tmpIndex + 1) % this.numberOfCard
        } while (this.showed[tmpIndex])
        return tmpIndex
    },

    hideCard: function (index) {
        this.showed[index] = false
    }
})